### ISSUE TITLE
Is there any workaround for issue #26703

### ISSUE THREAD

**Issue by SoftCircuits** [dotnet/efcore#32046]
I have the following query.

    var railcar = await DbContext.Railcars
        .Where(r => r.Id == id && r.FacilityId == UserContext.FacilityId)
        .AsNoTracking()
        .Select(r => new RailcarDetailsModel(facilityTime)
        {
            RailcarNumber = r.RailcarNumber,
            Track = r.Track.Name,
            Arrival = r.Arrival,
            Departure = r.Departure,
            Product = r.Product.Name,
            Consignee = r.Consignee,
            Shipper = r.Shipper,
            Origin = r.Origin,
            OriginalQuantity = r.InboundQuantity,
            Customer = r.Customer.Name,

            Trucks = r.ToTransfers
                .Where(t => t.FromTruckId.HasValue)
                .Select(t => new RailcarTruckModel
                {
                    Id = t.FromTruckId.Value,
                    TruckCompany = t.FromTruck.TruckCompany,
                    TruckNumber = t.FromTruck.TruckNumber,
                    Arrival = t.FromTruck.Arrival,
                    Departure = t.FromTruck.Departure,
                    BillOfLading = t.FromTruck.BillOfLading,
                    ManualEntry = t.FromTruck.ManualEntry,
                    Volume = t.Quantity
                })
                .Concat(r.FromTransfers
                    .Where(t => t.ToTruckId.HasValue)
                    .Select(t => new RailcarTruckModel
                    {
                        Id = t.ToTruckId.Value,
                        TruckCompany = t.ToTruck.TruckCompany,
                        TruckNumber = t.ToTruck.TruckNumber,
                        Arrival = t.ToTruck.Arrival,
                        Departure = t.ToTruck.Departure,
                        BillOfLading = t.ToTruck.BillOfLading,
                        ManualEntry = t.ToTruck.ManualEntry,
                        Volume = -t.Quantity
                    }))
                .OrderBy(rt => rt.Arrival)
                .ToList(),

            Storage = r.ToTransfers
                .Where(t => t.FromStorageId.HasValue)
                .Select(t => new RailcarStorageModel
                {
                    Id = t.Id,
                    TimeStamp = t.TimeStamp,
                    StorageId = t.FromStorageId.Value,
                    Storage = t.FromStorage.Name,
                    Volume = t.Quantity
                })
                .Concat(r.FromTransfers
                    .Where(t => t.ToStorageId.HasValue)
                    .Select(t => new RailcarStorageModel
                    {
                        Id = t.Id,
                        TimeStamp = t.TimeStamp,
                        StorageId = t.ToStorageId.Value,
                        Storage = t.ToStorage.Name,
                        Volume = -t.Quantity
                    }))
                .OrderBy(rs => rs.TimeStamp)
                .ToList(),
        })
        .FirstOrDefaultAsync();

When executed, it throws an exception.

> System.InvalidOperationException: 'Unable to translate a collection subquery in a projection since either parent or the subquery doesn't project necessary information required to uniquely identify it and correctly generate results on the client side. This can happen when trying to correlate on keyless entity type. This can also happen for some cases of projection before 'Distinct' or some shapes of grouping key in case of 'GroupBy'. These should either contain all key properties of the entity that the operation is applied on, or only contain simple property access expressions.'

If I change the two `Concat()`s to `Union()`s, it works okay. But I think I need `Concat()` to ensure all rows are included, duplicate or not.

I understand this is related to issue #26703, and is related to the fact that after my `Select()` statements, it is no longer to uniquely identify some of the rows.

But what is the workaround? I added the primary key (`Id`) to the projection but that didn't help. Is there any workaround for this case?

**Comment by ajcvickers** [comment]
This issue is lacking enough information for us to be able to fully understand what is happening. Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by maumar** [comment]
@SoftCircuits you can project the components of set operations separately and concat them on the client. Not a great workaround, but should work. Something like this:

```cs
var query = await DbContext.Railcars
    .Where(r => r.Id == id && r.FacilityId == UserContext.FacilityId)
    .AsNoTracking()
    .Select(r => new 
    {
        RailcarNumber = r.RailcarNumber,
        Track = r.Track.Name,
        Arrival = r.Arrival,
        Departure = r.Departure,
        Product = r.Product.Name,
        Consignee = r.Consignee,
        Shipper = r.Shipper,
        Origin = r.Origin,
        OriginalQuantity = r.InboundQuantity,
        Customer = r.Customer.Name,

        Trucks1 = r.ToTransfers
            .Where(t => t.FromTruckId.HasValue)
            .Select(t => new RailcarTruckModel
            {
                Id = t.FromTruckId.Value,
                TruckCompany = t.FromTruck.TruckCompany,
                TruckNumber = t.FromTruck.TruckNumber,
                Arrival = t.FromTruck.Arrival,
                Departure = t.FromTruck.Departure,
                BillOfLading = t.FromTruck.BillOfLading,
                ManualEntry = t.FromTruck.ManualEntry,
                Volume = t.Quantity
            }).ToList(),
        Trucks2 = r.FromTransfers
                .Where(t => t.ToTruckId.HasValue)
                .Select(t => new RailcarTruckModel
                {
                    Id = t.ToTruckId.Value,
                    TruckCompany = t.ToTruck.TruckCompany,
                    TruckNumber = t.ToTruck.TruckNumber,
                    Arrival = t.ToTruck.Arrival,
                    Departure = t.ToTruck.Departure,
                    BillOfLading = t.ToTruck.BillOfLading,
                    ManualEntry = t.ToTruck.ManualEntry,
                    Volume = -t.Quantity
                }).ToList();


)
            .OrderBy(rt => rt.Arrival)
            .ToList(),

        Storage1 = r.ToTransfers
            .Where(t => t.FromStorageId.HasValue)
            .Select(t => new RailcarStorageModel
            {
                Id = t.Id,
                TimeStamp = t.TimeStamp,
                StorageId = t.FromStorageId.Value,
                Storage = t.FromStorage.Name,
                Volume = t.Quantity
            }).ToList(),
        Storage2 = r.FromTransfers
                .Where(t => t.ToStorageId.HasValue)
                .Select(t => new RailcarStorageModel
                {
                    Id = t.Id,
                    TimeStamp = t.TimeStamp,
                    StorageId = t.ToStorageId.Value,
                    Storage = t.ToStorage.Name,
                    Volume = -t.Quantity
                }).ToList()
//            .OrderBy(rs => rs.TimeStamp)
//            .ToList(),
    })
    .FirstOrDefaultAsync();

var railcar = RailcarDetailsModel(facilityTime)
{
        RailcarNumber = query .RailcarNumber,
        Track = query.Track,
        Arrival = query .Arrival,
        Departure = query .Departure,
        Product = query .Product,
        Consignee = query .Consignee,
        Shipper = query.Shipper,
        Origin = query .Origin,
        OriginalQuantity = query .OriginalQuantity,
        Customer = query.Customer,

        Trucks = query.Trucks1
            .Concat(query.Trucks2)
            .OrderBy(rt => rt.Arrival)
            .ToList(),

        Storage = query.Storage1
            .Concat(query.Storage2)
            .OrderBy(rs => rs.TimeStamp)
            .ToList(),
};
```

**Comment by SoftCircuits** [comment]
Yes, processing on the client is always an option. In this case, it doesn't incur too much additional overhead. But it still seems like I should be able to do this.

**Comment by maumar** [comment]
It may look trivial on the surface but it clashes with one of the fundamental design assumptions we have in EF query and is really hard to fix. This has nothing to do with Tracking/NoTracking btw - even when we don't track the results we still need to reason about their relation to each other, so that we can bucket them accordingly in the final result. 

We have improved it in the past (fixing the Distinct case) and we have some ideas to improve other problematic cases, e.g. by introducing row number. We may try to improve it in the future versions, but for now the workaround is all we have.

**Comment by SoftCircuits** [comment]
@maumar I appreciate that. I won't pretend that I understand the issue with the information I have. But can you help me understand why it works fine with `Union()` but not `Concat()`?

**Comment by maumar** [comment]
union is basically Concat + Distinct. In case of Distinct, we can use the entire projection as the identifier (since Distinct guarantees the projection, when looked at in it's entirety, to be unique) This was one of the improvements in that area I was talking about. In case of Concat we can't do that, because there can be duplicates.

**Comment by SoftCircuits** [comment]
@ajcvickers I've completed a stand-alone demo of this issue and am attaching it here.

Maumar was kind enough to describe the issue and it appears a workaround is unlikely. Nonetheless, I'm following up with a demo. While I understand the core issue here, it will still be a big deal if a workaround could be devised.

[NoTranslareQueryExample.zip](https://github.com/dotnet/efcore/files/13032898/NoTranslareQueryExample.zip)

**Comment by ajcvickers** [comment]
@SoftCircuits The workaround here is to perform the concat on the client. This is two round-trips, but doesn't bring back any additional data. This is the best EF Core can do right now.

**Comment by owareoro** [comment]
For anyone who is still facing this issue - it is mentioned here although not explicitly that there is in fact a work-around to this issue.

At the expense of a less efficient generated SQL, you can fix this error by appending Distinct to all branches of your query that EF may consider to be non-distinct.

These are basically, queries with includes, selected projections and joins.

Rather than,

`query1.Concat(query2)`, you can work with
`query1.Distinct().Union(query2)` (or `query1.Distinct().Union(query2.Distinct())` if both fail the criteria for being unique).

This works very fine - and the only thing missing is a method that says (AssumeDistinct()) so we can skip the Distinct on the individual queries (and possibly on Concat) when we are sure the result is in fact distinct.

I also understand that the EF team knows best as to what features to add - this workaround works fine enough for me, at least for now.
